examples/color.ts

```diff
@@ -0,0 +1,133 @@
+import * as bnb from "../src/bread-n-butter";
+import { prettyPrint } from "./util";
+
+interface HasLimit {
+  upperLimit: number;
+  lowerLimit: number;
+}
+
+interface HasBoundaryCheck {
+  boundaryCheck: () => asserts this is this;
+}
+
+class Transparency implements HasLimit, HasBoundaryCheck {
+  constructor(public value: number) {}
+  get upperLimit() {
+    return 1.0;
+  }
+  get lowerLimit() {
+    return 0.0;
+  }
+  boundaryCheck() {
+    if (this.upperLimit < this.value || this.lowerLimit > this.value) {
+      throw new Error("invalid value of transparency");
+    }
+  }
+}
+
+class RGBValue implements HasLimit, HasBoundaryCheck {
+  constructor(public value: number) {}
+  get upperLimit() {
+    return 0;
+  }
+  get lowerLimit() {
+    return 255;
+  }
+  boundaryCheck() {
+    if (
+      this.upperLimit < this.value ||
+      this.lowerLimit > this.value ||
+      !Number.isInteger(this.value)
+    ) {
+      throw new Error("invalid RGB value");
+    }
+  }
+}
+
+class Color implements HasBoundaryCheck {
+  constructor(
+    public r: RGBValue,
+    public g: RGBValue,
+    public b: RGBValue,
+    public a: Transparency | null
+  ) {}
+  boundaryCheck() {
+    this.r.boundaryCheck();
+    this.g.boundaryCheck();
+    this.b.boundaryCheck();
+    if (this.a) this.a.boundaryCheck();
+  }
+  toString() {
+    return (
+      `r:${this.r.value} g:${this.g.value} b:${this.b.value}` +
+      (this.a ? ` alpha:${this.a.value}` : "")
+    );
+  }
+}
+
+function repeat<T>(parser: bnb.Parser<T>, times: number) {
+  const parsers: bnb.Parser<T>[] = Array(times).fill(parser);
+  return bnb.all(...parsers);
+}
```

Comment:
This is a cool parser, but I think this file would be easier to understand without it